xayn-ai/src/coi/system.rs

```diff
@@ -214,6 +223,160 @@ impl CoiSystem {
             })
             .collect()
     }
+
+    /// Selects the most relevant key phrases for the coi.
+    ///
+    /// The most relevant key phrases are selected from the set of key phrases of the coi and the
+    /// candidates. The computed relevances are a relative score from the interval `[0, 1]`.
+    #[allow(dead_code)]
+    fn select_key_phrases<
+        CP: CoiPoint + CoiPointKeyPhrases,
+        F: Fn(&str) -> Result<Embedding, Error>,
+    >(
+        &self,
+        coi: &mut CP,
+        candidates: &[String],
+        // TODO: make SMBert available to CoiSystem and remove this argument
+        smbert: F,
+    ) {
+        /// Reduces the matrix along the axis while skipping the diagonal elements.
+        fn reduce_without_diag<S, F, G>(a: ArrayBase<S, Ix2>, axis: Axis, f: F, g: G) -> Array2<f32>
+        where
+            S: Data<Elem = f32>,
+            F: Fn(f32, f32) -> f32,
+            G: Fn(f32) -> f32,
+        {
+            a.lanes(axis)
+                .into_iter()
+                .enumerate()
+                .map(|(i, lane)| {
+                    lane.iter()
+                        .enumerate()
+                        .filter_map(|(j, element)| (i != j).then(|| *element))
+                        .reduce(|reduced, element| f(reduced, element))
+                        .map(|reduced| g(reduced))
+                        .unwrap_or_default()
+                })
+                .collect::<Array1<_>>()
+                .insert_axis(axis)
+        }
+
+        /// Gets the index of the maximum element.
+        fn argmax<I, F>(iter: I) -> Ix
+        where
+            I: IntoIterator<Item = F>,
+            F: Borrow<f32>,
+        {
+            iter.into_iter()
+                .enumerate()
+                .reduce(|(arg, max), (index, element)| {
+                    if element.borrow() > max.borrow() {
+                        (index, element)
+                    } else {
+                        (arg, max)
+                    }
+                })
+                .map(|(arg, _)| arg)
+                .unwrap(/* at least one key phrase is selected */)
```

Comment:
it's correct that this is not guaranteed by this specific function, but this function is private within another function, which essentially applies the guarantees of the outer function for the inner function as well. hence i guess it's ok to unwrap here, otherwise we would need to introduce some error handling or return an option, because a default value doesn't seem reasonable in this case.